noir-projects/aztec-nr/aztec/src/state_vars/shared_mutable/scheduled_delay_change.nr

```diff
@@ -0,0 +1,487 @@
+// delay to be decreased immediately then it'd be possible for the state variable to schedule a value change with a 
+// reduced delay, invalidating prior private reads.
+struct ScheduledDelayChange<INITIAL_DELAY> {
+    // Both pre and post are stored in public storage, so by default they are zeroed. By wrapping them in an Option, 
+    // they default to Option::none(), which we detect and replace with INITIAL_DELAY. The end result is that a
+    // ScheduledDelayChange that has not been initialized has a delay equal to INITIAL_DELAY, which is the desired
+    // effect. Once initialized, the Option will never be none again.
+    pre: Option<u32>,
+    post: Option<u32>,
+    block_of_change: u32,
+    // The _dummy variable forces INITIAL_DELAY to be interpreted as a numberic value. This is a workaround to
+    // https://github.com/noir-lang/noir/issues/4633. Remove once resolved.
+    _dummy: [Field; INITIAL_DELAY],
+}
+
+impl<INITIAL_DELAY> ScheduledDelayChange<INITIAL_DELAY> {
+    pub fn new(pre: Option<u32>, post: Option<u32>, block_of_change: u32) -> Self {
+        Self { pre, post, block_of_change, _dummy: [0; INITIAL_DELAY] }
+    }
+
+    /// Returns the current value of the delay stored in the data structure.
+    /// This function only returns a meaningful value when called in public with the current block number - for
+    //// historical private reads use `get_effective_minimum_delay_at` instead.
+    pub fn get_current(self, current_block_number: u32) -> u32 {
+        // The post value becomes the current one at the block of change, so any transaction that is included in the
+        // block of change will use the post value.
+
+        if current_block_number < self.block_of_change {
+            self.pre.unwrap_or(INITIAL_DELAY)
+        } else {
+            self.post.unwrap_or(INITIAL_DELAY)
+        }
+    }
+
+    /// Returns the scheduled change, i.e. the post-change delay and the block at which it will become the current
+    /// delay. Note that this block may be in the past if the change has already taken place.
+    /// Additionally, further changes might be later scheduled, potentially canceling the one returned by this function.
+    pub fn get_scheduled(self) -> (u32, u32) {
+        (self.post.unwrap_or(INITIAL_DELAY), self.block_of_change)
+    }
+
+    /// Mutates the delay change by scheduling a change at the current block number. This function is only meaningful
+    /// when called in public with the current block number.
+    /// The block at which the new delay will become effective is determined automatically:
+    ///  - when increasing the delay, the change is effective immediately
+    ///  - when reducing the delay, the change will take effect after a delay equal to the difference between old and
+    ///    new delay. For example, if reducing from 3 days to 1 day, the reduction will be scheduled to happen after 2
```

Comment:
I'm not sure if you're asking 'why are we waiting the minimum amount of time instead of always simply waiting a longer time?', or if you're not sure _why_ this is the correct amount of time to wait.

Reply:
Kinda both actually. When I read:
```
when reducing the delay, the change will take effect after a delay equal to the difference between old and new delay
```
It is not instantly clear to me why we use that and why that is the shortest time (i might be idiot) and I had to think about it 🤯.

Reply:
From past experience it can be very annoying to be forced to wait extra time when you don't need to (e.g. if decreasing a delay from a month to two weeks, having to wait a total of 4 vs 6 weeks is a big difference). Hopefully most people will be relatively satisfied with the behavior and not be nerd-sniped by the API reference?

Reply:
Sure, I get why, was just not something that instantly clicked for me just from the comment. But think that might just be me reading comprehension 😆